pkg/controller/hyperconverged/hyperconverged_controller.go

```diff
@@ -843,6 +852,145 @@ func (r *ReconcileHyperConverged) recoverHCOVersion(request *common.HcoRequest)
+		req.Logger.Info("failed to create KubeVirt configmap", "error", err.Error())
+		return false, err
+	}
+
+	backupCm := makeCmBackup(cm)
+
+	req.Logger.Info("creating KubeVirt configmap backup")
+	if err := r.client.Create(req.Ctx, backupCm); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			req.Logger.Info("KubeVirt configmap backup already exists")
+		} else {
+			req.Logger.Info("failed to create KubeVirt configmap backup", "error", err.Error())
+			return false, err
+		}
+	} else {
+		r.eventEmitter.EmitEvent(req.Instance, corev1.EventTypeNormal, "Created", fmt.Sprintf("Created ConfigMap %s", backupKvCmName))
+	}
+
+	modified := adoptOldKvConfigs(req, cm)
+
+	err := r.removeKvConfigMap(req, cm)
+	if err != nil {
+		return false, err
+	}
+
+	return modified, nil
+}
+
+func (r *ReconcileHyperConverged) removeKvConfigMap(req *common.HcoRequest, cm *corev1.ConfigMap) error {
+	req.Logger.Info("removing the kubevirt configMap")
+	err := hcoutil.ComponentResourceRemoval(req.Ctx, r.client, cm, req.Name, req.Logger, false, true)
+	if err != nil {
+		return err
+	}
+
+	r.eventEmitter.EmitEvent(req.Instance, corev1.EventTypeNormal, "Killing", fmt.Sprintf("Removed ConfigMap %s", kvCmName))
+
+	refs := make([]corev1.ObjectReference, 0, len(req.Instance.Status.RelatedObjects))
+	for _, obj := range req.Instance.Status.RelatedObjects {
+		if obj.Kind == "ConfigMap" && obj.Name == kvCmName {
+			continue
+		}
+		refs = append(refs, obj)
+	}
+
+	req.Instance.Status.RelatedObjects = refs
+	req.StatusDirty = true
+
+	return nil
+}
+
+func makeCmBackup(cm *corev1.ConfigMap) *corev1.ConfigMap {
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      backupKvCmName,
+			Namespace: cm.Namespace,
+			Labels:    cm.Labels,
+		},
+		Data: cm.Data,
+	}
+}
+
+// Read the old KubeVit configuration from the config map, and move them to the HyperConverged CR
+//
+// In case of wrong foramt of the configmap, the HCO ignores this error (but print it to the log) in order to prevent
+// an infinite loop (returning error will cause the same error again and again, and the only way to stop the loop
+// is to manually fix or delete the wrong configMap).
+func adoptOldKvConfigs(req *common.HcoRequest, cm *corev1.ConfigMap) bool {
+	modified := false
+	kvLiveMigrationConfig, ok := cm.Data[liveMigrationKey]
+	if !ok {
+		return false
+	}
+	hcoLiveMigrationConfig := hcov1beta1.LiveMigrationConfigurations{}
+	err := yaml.NewYAMLOrJSONDecoder(strings.NewReader(kvLiveMigrationConfig), 1024).Decode(&hcoLiveMigrationConfig)
+	if err != nil {
+		req.Logger.Error(err, "Failed to read the KubeVirt ConfigMap, and its content was ignored. This ConfigMap will be deleted. The backup ConfigMap called "+backupKvCmName)
+		return false
+	}
+
+	if !reflect.DeepEqual(req.Instance.Spec.LiveMigrationConfig, hcoLiveMigrationConfig) {
+		req.Logger.Info("updating the HyperConverged CR from the KeubeVirt configMap")
+		kvConfigMapToHyperConvergedCr(req, hcoLiveMigrationConfig)
```

Comment:
Why do we need this? Why don't we directly assign
`req.Instance.Spec.LiveMigrationConfig = hcoLiveMigrationConfig`

I am sure I am missing something but I don't know what it is :D

Reply:
That allow keeping HCO defaults if the configmap does not include all the keys.

Reply:
After reading the test cases carefully, I understood it. Thanks!